qa/tasks/cephfs/test_mirroring.py

```diff
@@ -509,9 +511,13 @@ def test_cephfs_mirror_blocklist(self):
         # check if the rados addr is blocklisted
         self.assertTrue(self.mds_cluster.is_addr_blocklisted(rados_inst))
 
-        # wait enough so that the mirror daemon restarts blocklisted instances
-        time.sleep(40)
-        rados_inst_new = self.get_mirror_rados_addr(self.primary_fs_name, self.primary_fs_id)
+        # get the new rados_inst
+        rados_inst_new = ""
+        with safe_while(sleep=1, tries=40, action='wait for mirror status rados_inst') as proceed:
+            while proceed():
+                rados_inst_new = self.get_mirror_rados_addr(self.primary_fs_name, self.primary_fs_id)
```

Comment:
This check fails in run: /a/vshankar-2024-05-06_12:08:33-fs-wip-vshankar-testing-20240506.064357-testing-default-smithi/7693889

The mirror daemon did not restart the blocklisted instance since `fs mirror status` still showed "blocklisted":

```
2024-05-07T19:19:22.683 DEBUG:teuthology.orchestra.run.smithi179:mirror status for fs: cephfs> ceph --admin-daemon /var/run/ceph/cephfs-mirror.asok fs mirror status cephfs@12
2024-05-07T19:19:22.811 INFO:teuthology.orchestra.run.smithi179.stdout:{
2024-05-07T19:19:22.811 INFO:teuthology.orchestra.run.smithi179.stdout:    "state": "blocklisted"
2024-05-07T19:19:22.811 INFO:teuthology.orchestra.run.smithi179.stdout:}
2024-05-07T19:19:22.825 DEBUG:tasks.cephfs.test_mirroring:command returned={
    "state": "blocklisted"
}
```

@joscollin please check the mirror daemon logs as in why the blocklisted instance wasn't restarted.

Reply:
@vshankar 
Not much showing in the mirror daemon logs `ceph-client.mirror.32194.log`.
Remember you've asked to remove `time.sleep(40)`? So may be, the 40 seconds in `safe_while` is not enough for the restart completion and `mirror status` to report the new rados_inst?

Reply:
Could you check the mirror daemon to see the restart interval?

Reply:
`cephfs_mirror_restart_mirror_on_blocklist_interval ` is 30 sec by default. But looks like it never entered this block it seems https://github.com/ceph/ceph/blob/main/src/tools/cephfs_mirror/Mirror.cc#L567-#L572 for doing a restart. I couldn't find that log text. May be, give more logs and rerun the test to see what happened?

Reply:
The logs would already be available in the failed run. Could you please check why that block wasn't executed?

Reply:
After scheduling update_fs_mirrors a few times, the service daemon removes the filesystem, shuts down FSMirror and unregister the watcher. 

In logs, we don't know the values of `failed_restart` and `blocklisted_restart` to check what happened. They both should be 0 obviously.

[This](https://github.com/ceph/ceph/blob/main/src/tools/cephfs_mirror/Mirror.cc#L565) check might be false too.

Reply:
@vshankar Where is this blocklist timeout configured? What's the default value? 
https://github.com/ceph/ceph/blob/main/qa/tasks/cephfs/test_mirroring.py#L502

Reply:
Update:

I've generated some debug logs by doing another run in your branch wip-vshankar-testing-20240506.064357. It shows that `failed_restart` and  `blocklisted_restart` are always 0. That's why the block https://github.com/ceph/ceph/blob/main/src/tools/cephfs_mirror/Mirror.cc#L567-#L572 is not getting executed. Checking more on this.

```
ceph-client.mirror.31062.log:2024-05-17T03:07:35.857+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:37.893+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:39.892+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:41.892+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:43.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:45.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:47.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:49.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:51.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:53.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:55.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:57.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:07:59.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:08:01.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:08:03.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:08:05.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:08:07.896+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:08:09.900+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=2, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:17.902+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:19.942+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:21.942+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:23.942+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:25.946+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:27.946+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:29.946+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:31.946+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:33.946+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:35.946+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:37.946+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:39.946+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:09:41.946+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=4, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:11:31.952+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=8, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:11:33.992+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=8, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:11:35.992+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=8, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:11:37.992+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=8, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:11:39.992+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=8, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:11:41.992+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=8, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:11:43.992+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=8, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:12:42.003+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=10, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:12:44.043+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=10, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:12:46.043+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=10, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:12:48.043+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=10, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:12:50.043+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=10, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:12:52.043+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=10, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:12:54.043+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=10, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:13:58.046+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:00.086+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:02.086+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:04.086+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:06.090+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:08.090+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:10.090+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:12.114+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:14.114+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:16.114+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:18.114+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:14:59.569+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:01.569+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:03.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:05.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:07.569+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:09.569+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:11.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:13.569+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:15.569+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:17.569+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:19.569+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:21.569+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:23.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:25.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:27.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:29.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:31.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:33.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:35.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:37.577+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:39.573+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:41.572+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:43.576+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:45.572+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:47.572+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:49.572+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 0is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:51.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:53.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:55.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:57.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:15:59.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:01.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:03.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:05.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:07.584+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:09.584+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:11.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:13.580+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:15.584+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:17.584+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0
ceph-client.mirror.31062.log:2024-05-17T03:16:19.584+0000 7fccb1519640  5 cephfs::mirror::Mirror update_fs_mirrors: filesystem={fscid=12, fs_name=cephfs} failed mirroring (failed: 0, blocklisted: 0), mirror_action.action_in_progress: 1is_restarting: 0

```

Reply:
> @vshankar Where is this blocklist timeout configured? What's the default value?
> https://github.com/ceph/ceph/blob/main/qa/tasks/cephfs/test_mirroring.py#L502

You should be able to find this in Mirror.cc and is a config value.

Reply:
> I've generated some debug logs by doing another run in your branch wip-vshankar-testing-20240506.064357. It shows that `failed_restart` and `blocklisted_restart` are always 0. That's why the block https://github.com/ceph/ceph/blob/main/src/tools/cephfs_mirror/Mirror.cc#L567-#L572 is not getting executed. Checking more on this.

You are on the right track. Figuring that bit out will RCA the bug.

Reply:
It's hard to print all the values involved in determining `failed_restart` and `blocklisted_restart`. But there's a possibility that `mirror_action.fs_mirror->get_blocklisted_ts()` returns `clock::now()` and then `blocklisted_restart` becomes false. I'll check that.

Reply:
@vshankar I have reproduced this issue on `wip-vshankar-testing-20240506.064357`. The culprit is the [duration check here](https://github.com/ceph/ceph/blob/main/src/tools/cephfs_mirror/Mirror.cc#L563). When the daemon in blocklisted, `get_blocklisted_ts()` returns `clock::now()` because `m_instance_watcher` and `m_mirror_watcher` check doesn't become `true` inside `get_blocklisted_ts()`. So we should use something smaller than `duration_cast<seconds>`, (may be milliseconds) to get a time diff. Or are they (`m_instance_watcher` and `m_mirror_watcher`) expected to be null during blocklisted? The problem lies there, why are they null during blocklist?

Reply:
Good catch. The correct approach would be to have FSMirror register a listener with InstanceWatcher/MirrorWatcher which would get invoked when the mirror daemon is blocklisted (or failed). That way FSMirror can maintain the last blocklisted timestamp and use that in get_blocklisted_ts().

Reply:
@vshankar 
That's good idea. How about checking if the daemon is blocklisted at the beginning of `schedule_mirror_update_task()`, iterating `m_mirror_actions` and then create a new `m_instance_watcher` via FSMirror (`mirror_action.fs_mirror->create_new_watcher_instance()`) ?

Reply:
> @vshankar That's good idea. How about checking if the daemon is blocklisted at the beginning of `schedule_mirror_update_task()`, iterating `m_mirror_actions` and then create a new `m_instance_watcher` via FSMirror (`mirror_action.fs_mirror->create_new_watcher_instance()`) ?

The same problem exist since `m_instance_watcher` can be `nullptr`, isn't it?

Reply:
Yes, both `m_instance_watcher` and `m_mirror_watcher` could be nullptr when blocklisted. Do you mean the same issue hits even after creating new watcher instance from `schedule_mirror_update_task` ?

Reply:
No. If rewatch fails due to blocklisting, then how does one check the blocklisted timestamp in the watchers? So, a listener would record that timestamp in FSMirror.

Reply:
@vshankar The PR is updated as per your suggestion. Created a new Listener and the InstanceWatcher and MirrorWatcher updates the blocklisted timestamp. But still the test fails if we have this check in Mirror.cc: `duration_cast<seconds>(mirror_action.fs_mirror->get_blocklisted_ts() - clock::now()) > blocklist_interval`.

Reply:
@vshankar Updated the fix. Will check whether the qa changes are needed by tomorrow.